src/irr/asset/CGLSLVirtualTexturingBuiltinIncludeLoader.h

```diff
@@ -0,0 +1,368 @@
+		for (uint32_t i = 0u; i < get_storage_cnt(); ++i)
+		{
+			uint32_t xbits = std::atoi(args[get_ix_addr_xmask_i(i)].c_str());
+			xbits = core::findMSB(xbits)+1;
+			s += std::to_string(xbits) + ((i==(get_storage_cnt()-1u)) ? "" : ", ");
+		}
+		s += ");";
+		s += "\nconst uint addr_layer_shift[] = uint[" + std::to_string(get_storage_cnt()) + "] (";
+		for (uint32_t i = 0u; i < get_storage_cnt(); ++i)
+		{
+			uint32_t xbits = std::atoi(args[get_ix_addr_xmask_i(i)].c_str());
+			xbits = core::findMSB(xbits)+1;
+			uint32_t ybits = std::atoi(args[get_ix_addr_ymask_i(i)].c_str());
+			ybits = core::findMSB(ybits)+1;
+			s += std::to_string(xbits+ybits) + ((i==(get_storage_cnt()-1u)) ? "" : ", ");
+		}
+		s += ");";
+		s += "\n\n#define PAGE_SZ " + std::to_string(1u<<pg_sz_log2) + "u" +
+			"\n#define PAGE_SZ_LOG2 " + args[ix_pg_sz_log2] + "u" +
+			"\n#define TILE_PADDING " + args[ix_tile_padding] + "u" +
+			"\n#define PADDED_TILE_SIZE uint(PAGE_SZ+2*TILE_PADDING)" +
+			"\n\nconst vec2 packingOffsets[] = vec2[PAGE_SZ_LOG2+1]( vec2(0.0,0.0)," + tilePackingOffsetsStr() + ");";
+		s+=
+R"(
+#define irr_glsl_STextureData_WRAP_REPEAT 0u
+#define irr_glsl_STextureData_WRAP_CLAMP 1u
+#define irr_glsl_STextureData_WRAP_MIRROR 2u
+
+vec3 irr_glsl_unpackPageID(in uint pageID, in uint formatID)
+{
+	// this is optimal, don't touch
+	uvec2 pageXY = uvec2(pageID,pageID>>addr_y_shift[formatID])&uvec2(addr_x_mask[formatID],addr_y_mask[formatID]);
+	return vec3(vec2(pageXY),float(pageID>>addr_layer_shift[formatID]));
+}
+uvec2 irr_glsl_unpackWrapModes(in uvec2 texData)
+{
+    return (texData>>uvec2(28u,30u)) & uvec2(0x03u);
+}
+uint irr_glsl_unpackMaxMipInVT(in uvec2 texData)
+{
+    return (texData.y>>24)&0x0fu;
+}
+vec3 irr_glsl_unpackVirtualUV(in uvec2 texData)
+{
+	// assert that PAGE_SZ_LOG2<8 , or change the line to uvec3(texData.yy<<uvec2(PAGE_SZ_LOG2,PAGE_SZ_LOG2-8u),texData.y>>16u)
+    uvec3 unnormCoords = uvec3(texData.y<<PAGE_SZ_LOG2,texData.yy>>uvec2(8u-PAGE_SZ_LOG2,16u))&uvec3(uvec2(0xffu)<<PAGE_SZ_LOG2,0xffu);
+    return vec3(unnormCoords);
+}
+vec2 irr_glsl_unpackSize(in uvec2 texData)
+{
+	return vec2(texData.x&0xffffu,texData.x>>16u);
+}
+)";
+		s += //those are supposed to be used by VT fucntions only (hence extra VT_ pefix)
+			"\n#define irr_glsl_VT_getPgTabSzLog2 " + args[ix_get_pgtab_sz_log2_name] +
+			"\n#define irr_glsl_VT_getPhysPgTexSzRcp " + args[ix_get_phys_pg_tex_sz_rcp_name] +
+			"\n#define irr_glsl_VT_getVTexSzRcp " + args[ix_get_vtex_sz_rcp_name] +
+			"\n#define irr_glsl_VT_layer2pid " + args[ix_get_layer2pid_name]
+			;
+        s += R"(
+vec3 irr_glsl_vTextureGrad_helper(in uint formatID, in vec3 virtualUV, in int clippedLoD, in int levelInTail)
+{
+    uvec2 pageID = textureLod(pageTable,virtualUV,clippedLoD).xy;
+
+	const uint pageTableSizeLog2 = irr_glsl_VT_getPgTabSzLog2(formatID);
+    const float phys_pg_tex_sz_rcp = irr_glsl_VT_getPhysPgTexSzRcp(formatID);
```

Comment:
this might need to change, see https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&hide-deleted-files=true#r417432177

Reply:
why is github so fucking stupid, now I can't find what I was linking!